### PR TITLE
Retry modem initialization on failures

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

File: src/base/service.toit [hunk 1]
```diff
@@ -141,6 +146,11 @@ abstract class CellularServiceProvider extends ProxyingNetworkServiceProvider:
 
     driver/Cellular? := null
     catch: driver = open_driver logger
+    // If we failed to create the driver, it may very well be
```
Comment (JohanEThomsen) [inline src/base/service.toit:149]: Nice :)

**Comment by kasperl** [comment]
TBR'ed.

File: src/base/service.toit [hunk 2]
```diff
@@ -265,25 +277,25 @@ abstract class CellularServiceProvider extends ProxyingNetworkServiceProvider:
...
+      if reset == CELLULAR_RESET_SOFT:
         driver.reset
-        reset_ = CELLULAR_RESET_NONE
         sleep --ms=1_000
-      else if reset_ == CELLULAR_RESET_POWER_OFF:
+      else if reset == CELLULAR_RESET_POWER_OFF:
         driver.power_off
-        reset_ = CELLULAR_RESET_NONE
         sleep --ms=1_000
       with_timeout --ms=20_000: driver.wait_for_ready
```
Comment (floitsch) [inline src/base/service.toit:286]: The Sara 5 file says that a power off can take 25s.

File: src/base/service.toit [hunk 3]
```diff
@@ -265,25 +277,25 @@ abstract class CellularServiceProvider extends ProxyingNetworkServiceProvider:
...
         // Turning the cellular modem on failed, so we artificially
         // bump the number of attempts to get close to a reset.
         if attempts_until_reset > 1:
-          attempts_ += attempts_until_reset - 1
-          critical_do: bucket_[ATTEMPTS_KEY] = attempts_
+          critical_do: update_attempts_ attempts + attempts_until_reset - 1
+        // Close the UART before closing the pins. This is typically
+        // taken care of by a call to driver.close, but in this case
+        // we failed to produce a working driver instance.
+        port.close
```
Comment (floitsch) [inline src/base/service.toit:297]: This sounds/feels iffy. Works be cleaner if ownership was more consistent

**Review by floitsch — APPROVED** [review]
LGTM

File: src/base/service.toit (hunk 3 above)
Comment (kasperl) [reply src/base/service.toit:297]: I agree :(

File: src/base/service.toit (hunk 2 above)
Comment (kasperl) [reply src/base/service.toit:286]: We're waiting for the power off in `driver.power_off` too.

**Pull request merged**